components/search/src/filter.rs

```diff
@@ -108,11 +110,91 @@ fn maybe_extract_engine_config(
         .map(|_variant| SearchEngineDefinition::from_configuration_details(&identifier, base))
 }
 
+fn determine_default_engines(
+    engines: &[SearchEngineDefinition],
+    default_engines_record: Option<JSONDefaultEnginesRecord>,
+    user_environment: &SearchUserEnvironment,
+) -> (Option<String>, Option<String>) {
+    match default_engines_record {
+        None => (None, None),
+        Some(record) => {
+            let mut default_engine_id = None;
+            let mut default_engine_private_id = None;
+
+            let specific_default = record
+                .specific_defaults
+                .into_iter()
+                .rev()
+                .find(|r| matches_user_environment(&r.environment, user_environment));
+
+            if let Some(specific_default) = specific_default {
+                // Check the engine is present in the list of engines before
+                // we return it as default.
+                let engine_id = find_engine_with_match(engines, specific_default.default);
+                if engine_id.is_some() {
+                    default_engine_id = engine_id;
+                }
+                let private_engine_id =
+                    find_engine_with_match(engines, specific_default.default_private);
+                if private_engine_id.is_some() {
+                    default_engine_private_id = private_engine_id;
+                }
+            }
+
+            (
+                // If we haven't found a default engine in a specific default,
+                // then fall back to the global default engine - but only if that
+                // exists in the engine list.
+                //
+                // For the normal mode engine (`default_engine_id`), this would
+                // effectively be considered an error. However, we can't do anything
+                // sensible here, so we will return `None` to the application, and
+                // that can handle it.
+                default_engine_id.or_else(|| engine_id_if_found(engines, record.global_default)),
+                default_engine_private_id
+                    .or_else(|| engine_id_if_found(engines, record.global_default_private)),
```

Comment:
Two questions: 

1. I don't quite understand the second paragraph -- what is "normal mode engine"?  If it uses the global default as the default engine in this situation, why is that considered an error?  

2. Remind me again, how does the application layer handle `None` when we're in the situation with no default engine found?

Reply:
1. Normal mode = not in private browsing mode. It is considered an error if we don't have a default engine, as we should always have an application default engine specified.

2. It is up to the application layer. On desktop we fall back to the first "general" search engine found (iirc according to alphabetical order).

Reply:
Oh I see. Yeah for no.1 if we fall back to the global default, it means the user is in some kind of environment where it didn't match any engines in the config or something else went wrong, and no default appProvided search engine was found for them.